### ISSUE TITLE
`machine_learning/linear_regression.py` doesn't give optimal coefficients

### ISSUE THREAD

**Issue by tianyizheng02** [TheAlgorithms/Python#8847]
### Feature description

Related to issue #8844

TL;DR: the current implementation doesn't give optimal solutions, the current implementation calculates SSE wrong, and we should add an implementation of a numerical methods algorithm that actually gives optimal solutions

In `machine_learning/linear_regression.py`, add the following code at the bottom of the `main()` function:
```py
from sklearn.linear_model import LinearRegression
import matplotlib.pyplot as plt

data = np.asarray(data.astype(float))
X = data[:, 0].reshape(-1, 1)
y = data[:, 1]

reg = LinearRegression().fit(X, y)
print(f"Sklearn coefficients: {reg.intercept_}, {reg.coef_}")

sse = np.sum(np.square(reg.predict(X) - y.T))
print(f"{sse = }")
print(f"mse = {sse / len(y)}")
print(f"half mse = {sse / (2 * len(y))}")

plt.scatter(X, y, color="lightgray")
plt.axline(xy1=(0, theta[0, 0]), slope=theta[0, 1], color="red", label="Gradient descent")
plt.axline(xy1=(0, reg.intercept_), slope=reg.coef_, color="blue", label="Sklearn")
plt.legend()
plt.show()
```
This code performs ordinary least squares (OLS) linear regression using `sklearn` as a point of reference to compare the current implementation against. It then calculates the sum of squared errors (SSE), mean squared error (MSE), and half of the MSE. To compare the outputs visually, the code uses `matplotlib` to plot the `sklearn` regression line and the regression line produced by the current implementation.

The code produces the following command line output:
```
...
At Iteration 100000 - Error is 128.03882
Resultant Feature vector: 
-9.34325
1.53067
Sklearn coefficients: -15.547901662158367, [1.6076036]
sse = 253795.17406773588
mse = 253.79517406773587
half mse = 126.89758703386794
```
As we can see, what the implementation calculates as the SSE (128.03882) is actually half of the MSE, meaning that the `sum_of_square_error` function is incorrect and needs to be fixed. Why the implementation is using _half_ of the MSE, I have no clue.

Furthermore, we can see that both the regression coefficients and the errors are slightly off. This is because the current implementation works via gradient descent, meaning that it can only _approximate_ the OLS regression coefficients. Meanwhile, libraries like `numpy` and `sklearn` calculate the mathematically optimal coefficients using [numerical methods](https://en.wikipedia.org/wiki/Numerical_methods_for_linear_least_squares).
![linear_regression_plot](https://github.com/TheAlgorithms/Python/assets/52298854/a43d60bf-2c9f-41eb-b96a-94ff9c4f4fde)
Although using gradient descent to perform linear regression does work, it's still suboptimal and (AFAIK) it's not how linear regression is actually performed in practice. We can still include this implementation, but we should definitely also include an implementation of an optimal numerical method.

**Comment by Madhav-2808** [comment]
Assign this task to me :)

**Comment by tianyizheng02** [comment]
> Assign this task to me :)

@Madhav-2808 We don't assign tasks in this repo. If you want to work on it, then just make a PR.

**Comment by liakdimi1** [comment]
@tianyizheng02  I try to run your code and get your graph but i get the error 

ValueError: setting an array element with a sequence. The requested array has an inhomogeneous shape after 2 dimensions. The detected shape was (2, 2) + 
inhomogeneous part.

**Comment by tianyizheng02** [comment]
@liakdimi1 I'm not able to recreate your error. Can you show your code and/or the error traceback?

**Comment by williamsmith2626** [comment]
Absolutely! When it comes to achieving optimal results, precision matters just as much in coding as it does in [Absolute Thai menu](https://mlymenu.com/category/thai/). Just like how our chefs carefully balance flavors, debugging code ensures linear regression models yield perfect coefficients. Keep refining, just like we do with our Thai dishes!

**Comment by williamsmith2626** [comment]
Assign this task to me. Recently was busy in creating site  [Servicio de Enfermeria a Domicilio](https://www.enfermerasom.com/servicios/) Now free

**Comment by JrX970** [comment]
If the problem is just focussing on the univariate case (as plotted), wouldn't it be worthwhile ditching the iterative closed form solution method and just doing a simple "sum of rectangular area over sum of square area" method? 
Especially considering that you would have to manually edit the code to increase the amount of features used, applying any data transformation or altering the design matrix.

**Comment by mahi01agarwal** [comment]
Hey, I have created a PR #9114 which addresses this issue. Please review it. Thanks!

**Comment by ssom01-github** [comment]
@tianyizheng02, Hello, I can fix this issue in the best possible way. Kindly assign this issue to me

**Comment by tianyizheng02** [comment]
@mahi01agarwal @ssom01-github Please read the [contributing guidelines](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md#contribution). As it says, _we do not assign issues, so do not ask for permission to work on an issue._ If you want to contribute, just open a PR. However, at the moment, please wait until October 1 in your time zone to open a PR because we're currently not accepting new PRs ahead of Hacktoberfest.

**Comment by UjjwalPardeshi** [comment]
can you assign me this task please .

**Comment by Sarahsmith6y5ttt** [comment]
I am also facing the same issue for [Geometry dash](https://geometrydashapps.com.mx/) site.

**Comment by smruthi-sumanth** [comment]
HI! Can I work on this issue?

**Comment by smruthi-sumanth** [comment]
Hello! My PR #11311 attempts to rectify this issue by implementing OLS instead of gradient descent. Can someone please review my code?
